### ISSUE TITLE
questions that would be useful to me to see answered in the readme..

### ISSUE THREAD

**Issue by lecstor** [stampit-org/stampit#93]
I'm reading the docs and some of the blog posts and I love the concepts of simple functions and reusability. I've formed a few ideas on how I could build simple classes that import a lot of their functionality from one or more shared function libraries. Now I'm looking at stampit and wondering if it's overkill for what I want to do. _At this stage_ I only really need maybe a half dozen classes some very similar, some not so I'm wondering..

- when and how should it be used? 
ie.
- what benefits would it bring to a very simple object structure if thats all I need?
- should it be saved for when I need lots of objects that need to share common code and complex behaviours?

I'll figure these things out for myself eventually I'm sure but I'm always happy for a head start.

cheers

**Comment by lecstor** [comment]
and having a play for a couple of hours it's great and it doesn't need to be as complex as it seems the first time you read the readme.. I was so confident I thought I'd write a little example that might have helped me.. just writing this I learned a few more things, but I have to admit the init args still has me a bit perplexed..

```
import stampit from 'stampit'
import assign from 'lodash/object/assign'

let Thing = stampit({
  init: ({self, stamp, args}) => { assign(this, args) },
  methods: {
    myReverse(arg1, arg2) {
      return [arg2, arg1]
    }
  }
})

let OtherThing = Thing
  .init(
    function ({self, stamp, args}){
      console.log('init', args)
      this.arg2 = args
    }
  )
  .methods({
    myMultiplier(arg1, arg2) {
      return [arg1, arg2, arg1, arg2]
    }
  })

let thingInstance = OtherThing({ key: 'value' }, 'arg2')

console.log('instance', thingInstance)
console.log('reverse', thingInstance.myReverse('three','four'))
console.log('multiplier', thingInstance.myMultiplier('three','four'))

---
init [ 'arg2' ]
instance { key: 'value', arg2: [ 'arg2' ] }
reverse [ 'four', 'three' ]
multiplier [ 'three', 'four', 'three', 'four' ]
```

**Comment by ericelliott** [comment]
Writing stamps that take arguments is an anti-pattern. Stamps support arguments primarily to support legacy constructor arguments. When you're dealing with stamp composition, there is a high probability of argument collision which could create some very unexpected and frustrating results.

**Comment by ericelliott** [comment]
I wish I had time for more thorough and helpful feedback but I'm strapped for time today. Lucky, the other Stampit community members are amazing and wonderful people. I'm sure they'll be very helpful.

Welcome to the community. We're glad to have you! =)

**Comment by lecstor** [comment]
wow, ok, I totally missed that one. Thank you. I'm now looking at refs() and my mental model has rust flipped to where I'm attaching stamps to the nodes in my object (created by parsing an xmlish document).

**Comment by lecstor** [comment]
possibly a better example.. I think I know what I need to do (build up a tree of node objects) but it's not what I'd like to do (load a node into a stamp, access/modify and have modifications reflected in the main object).

```import stampit from 'stampit'
import map from 'lodash/collection/map'

let Node = stampit({
  methods: {
    getGutter(sides, types) {
      let gutter = 0
      map(types, (type) => {
        if (this.node.attr[type]){
          map(sides, (side) => { gutter += this.node.attr[type][side]-0 })
        }
      })
      return gutter
    },
    setPaddingLeft(size) {
      this.node.attr.padding.left = size
    }
  }
})

let doc = {
  kids: [
    {
      attr: {
        padding: { left: 5, right: 10 },
        margin: { left: 3, right: 6 }
      }
    }
  ]
}

let node = Node.create({ node: doc.kids[0] })

node.setPaddingLeft(18)

console.log('doc', JSON.stringify(doc, null, 2))
console.log('node', JSON.stringify(node.node, null, 2))
```

**Comment by ericelliott** [comment]
> @ericelliott no worries about the time thing, totally understand, thanks for responding. I'll keep playing and learning (learning loads every step of the way). @koresar asked me to share my thoughts and questions, so that's mainly all I'm doing.

Very cool. Thanks for taking the time. It really helps us understand how we can flesh out the documentation to better orient beginners to using stamps. =)

**Comment by koresar** [comment]
> So create() would seem to be the way, but it looks like that is doing a deepcopy

Nope, it copies by reference. Moreover, `Node.create()` and `Node()` is exactly the same call.
I believe we have that in the docs, but obviously that info is hidden in the pile of other "very useful" info.

I'm seeing people are getting very confused of our readme. :(

**Comment by lecstor** [comment]
so what am I doing wrong then, or how am I explaining myself incorrectly? 8)
Using refs() achieves my goal of setting the `padding.left` value in `doc`, but creating another stamp for each instance is not what I'd want to do.
```
import stampit from 'stampit'
import map from 'lodash/collection/map'

let Node = stampit({
  methods: {
    setPaddingLeft(size) {
      this.attr.padding.left = size
    }
  }
})

let doc = { kids: [ {attr: {padding: { left: 5, right: 10 } } } ] }

let node = Node(doc.kids[0])
node.setPaddingLeft(18)
console.log('doc', JSON.stringify(doc, null, 2))

node = Node.refs(doc.kids[0])()
node.setPaddingLeft(18)
console.log('refdoc', JSON.stringify(doc, null, 2))

>>>>

doc {
  "kids": [
    {
      "attr": {
        "padding": {
          "left": 5,
          "right": 10
        }
      }
    }
  ]
}
refdoc {
  "kids": [
    {
      "attr": {
        "padding": {
          "left": 18,
          "right": 10
        }
      }
    }
  ]
}

```

**Comment by troutowicz** [comment]
I think you are trying to do something like this?

```js
let doc = { kids: [ {attr: {padding: { left: 5, right: 10 } } } ] }

let Node = stampit()
  .refs(doc.kids[0])
  .methods({
    setPaddingLeft(size) {
      this.attr.padding.left = size;
    }
  });

let node = Node();

node.setPaddingLeft(18)
console.log(doc.kids[0].attr.padding.left);

node.setPaddingLeft(5)
console.log(doc.kids[0].attr.padding.left);
```

**Comment by lecstor** [comment]
yes, that's turning my last example around a bit. I achieves the result of updating the main doc by reference but involves creating a new stamp for each instance which doesn't seem right. I'll be creating hundreds of instances where only the kid is different.
Ideally I would be able to do
```
let node1 = Node(doc.kids[0])
let node2 = Node(doc.kids[1])
let node3 = Node(doc.kids[2])
node1.setPaddingLeft(18)
node2.setPaddingLeft(10)
node3.setPaddingLeft(15)
```
and have the doc object updated accordingly.

**Comment by troutowicz** [comment]
I didn't turn your example around, I corrected it.

In your example `let node = Node(doc.kids[0])` does not set `refs`. As @ericelliott said above, passing arguments is an anti-pattern and should be avoided.

You don't need to create a new instance for each node, just create one doc instance.

```js
let src = { kids: [ {attr: {padding: { left: 5, right: 10 } } } ] }

let Doc = stampit()
  .refs({ kids: src.kids })
  .methods({
    setPaddingLeft(node, size) {
      this.kids[node - 1].attr.padding.left = size;
    }
  });

let doc = Doc();

doc.setPaddingLeft(1, 18)
console.log(src.kids[0].attr.padding.left);

doc.setPaddingLeft(1, 5)
console.log(src.kids[0].attr.padding.left);
```

**Comment by lecstor** [comment]
sorry, I was referring to the last example in that post..
```
node = Node.refs(doc.kids[0])()
node.setPaddingLeft(18)
console.log('refdoc', JSON.stringify(doc, null, 2))
```
which I think does what you're doing there (with a bit of short-hand).

I'll try to give a better example, but keep it as concise as possible.
Here is an example doc for which I would like to use different "classes" to handle the different node types as each node type has different behaviour.
```
  let doc = {
    attr: {
      padding: { left: 5, right: 10 },
      margin: { left: 3, right: 6 }
    },
    kids: [
      {
        tag: 'row'
        attr: { padding: { left: 5, right: 10 }, },
        kids: [
          {
            tag: 'cell',
            attr: { padding: { left: 5, right: 10 }, },
            kids: 'Hello'
          },
          {
            tag: 'cell',
            attr: { padding: { left: 5, right: 10 }, },
            kids: 'Goodbye'
          }
        ]
      }
    ]
  }
```
so what I'm hearing is that to do what I want to do I need to create a new stamp for each node, then instanciate a single instance from that stamp which to me feels like creating a class & instance for each node.

**Comment by lecstor** [comment]
so, yeh, you're confirming what I said about `Node.create()` which is the same as `Node()` 
> So create() would seem to be the way, but it looks like that is doing a deepcopy

The notation `Node.refs({ ba: 'boom' })()` does what I want, but feels like it's doing extra work just to get the instance created to contain a reference to the main doc compared to `Node({ ba: 'boom' })`.

checking out for a bit, thanks for your help.

**Comment by troutowicz** [comment]
I'm not sure if you are understanding how an instance is created with stampit. In your previous example

```js
let node1 = Node(doc.kids[0])
let node2 = Node(doc.kids[1])
let node3 = Node(doc.kids[2])
node1.setPaddingLeft(18)
node2.setPaddingLeft(10)
node3.setPaddingLeft(15)
```
You just created 3 separate instances. stampit is a factory that returns a factory.

**Comment by lecstor** [comment]
Yeh, looks like I'm definitely missing something somewhere, will keep at it, thanks again.

**Comment by troutowicz** [comment]
You're welcome!

**Comment by unstoppablecarl** [comment]
@lecstor if you get on the gitter channel I can walk you through it https://gitter.im/ericelliott/stampit?utm_source=share-link&utm_medium=link&utm_campaign=share-link

**Comment by lecstor** [comment]
> stampit is a factory that returns a factory

But at some stage you need an object instance, my goal isn't to build factories, it's to do useful stuff 8)
And the idea of a class is that you use it to create object instances that have different properties but shared behaviour.

My conundrum here is that I wish for the properties of my objects to be references to a larger object.
To do that with stampit I need to use `refs()` which means creating another class with my refs attached and then instantiating an instance of that class rather than having a class that I can create all of my objects from. I could do this with Node.create() if I didn't want the properties to be a reference.

So if that's just something stampit wont do, then cool, fair enough. If it will, then I'm hoping someone can tell me how as I can't get it to comply.

cheers

**Comment by lecstor** [comment]
ok. the bottom line.. Is this considered a bad or a waste of cycles?
(no different to earlier code, but does what I want with a convenience method)

```
import stampit from 'stampit'
import map from 'lodash/collection/map'

let Node = stampit({
  methods: {
    setPaddingLeft(size) {
      this.attr.padding.left = size
    }
  },
  static: {
    fromRef: function(args) {
      return this.refs(args)()
    }
  }
})

let doc = { kids: [
  {attr: {padding: { left: 5, right: 10 } } },
  {attr: {padding: { left: 20, right: 20 } } }
]}

let node1 = Node.fromRef(doc.kids[0])
let node2 = Node.fromRef(doc.kids[1])
node1.setPaddingLeft(18)
node2.setPaddingLeft(16)
console.log('doc', JSON.stringify(doc, null, 2))
```

**Comment by troutowicz** [comment]
Yeah, you aren't understanding. Let's step through your code.

```js
let Node = stampit({ ... });
```

* You are still using args when we have said it's an anti-pattern.
* What do you think Node is at this point?
  * Hint 1: `console.log(typeof Node);`
  * Hint 2: it's not a stamp instance

```js
fromRef: function(args) {
  return this.refs(args)()
}
```
You have declared `fromRef` a `static` function. `this` will be equal to `Node`, which is not an object instance and does not have access to object instances and their state.

```js
let node1 = Node.fromRef(doc.kids[0])
```
What do you think this is doing? What it is actually doing is returning a shallow copy of `doc.kids[0]`. It is in no way related to Node (since node is a factory not an instance).

In the end all you did was go in a big circle, never actually making use of your stamp.

**Comment by lecstor** [comment]
8) hehe. I'm having trouble not believing what I'm seeing. It looks like a duck, and walks like a duck.. (and behaves like I want). I'm not sure how it's not making use of the stamp as it gives me an object which I can call methods on to manipulate the data I want it to.

Args are an anti-patter, but isn't that exactly what create(properties) does? I'm just working around the fact that create deepcopies the properties so I lose my reference.

What if I expand the example to it's intended usage?
```
import stampit from 'stampit'
import map from 'lodash/collection/map'

let Node = stampit({
  methods: {
    setPaddingLeft(size) {
      this.attr.padding.left = size
    }
  },
  static: {
    fromRef: function(args) {
      return this.refs(args)()
    }
  }
})

let Cell = Node.methods({
  ima(){
    console.log('== Ima Cell ==', this.attr.padding.left)
  }
})

let Row = Node.methods({
  ima(){
    console.log('== Ima Row ==', this.attr.padding.left)
  }
})


let doc = { kids: [
  {attr: {padding: { left: 5, right: 10 } } },
  {attr: {padding: { left: 25, right: 50 } } },
]}

let node1 = Cell.fromRef(doc.kids[0])
node1.setPaddingLeft(18)
node1.ima()

let node2 = Row.fromRef(doc.kids[1])
node2.setPaddingLeft(28)
node2.ima()

console.log('doc', JSON.stringify(doc, null, 2))

>>>

== Ima Cell == 18
== Ima Row == 28
doc {
  "kids": [
    {
      "attr": {
        "padding": {
          "left": 18,
          "right": 10
        }
      }
    },
    {
      "attr": {
        "padding": {
          "left": 28,
          "right": 50
        }
      }
    }
  ]
}

```

**Comment by koresar** [comment]
There are many wrong things in this thread. Let me get to a computer in few
hours. I'll explain where we were misguided.

**Comment by lecstor** [comment]
8) cool, back to my chores then.. have a good day @koresar!

**Comment by koresar** [comment]
I'll take a chance to answer later. For now I think @lecstor has found a *bug*.

Stampit v1 factories were mixin' (shallow copy) the first argument - `stamp(someRefs)`.
Stampit v2 factories are deep merging/cloning the first argument - `stamp(objToClone)`.

@ericelliott in this thread we decided to deep clone the factory argument: https://github.com/ericelliott/stampit/issues/40#issuecomment-70052358
Was that a correct decision back then?

IMO we should have two options:
`stamp(refs)` or `stamp.create(refs)` - should shallow copy.
`stamp.createFrom(props)` - should deep merge/clone. (Please, suggest better name.)
OR
`stamp(refs)` - for shallow copying.
`stamp.create(props)` - for deep merging/cloning.

**Comment by koresar** [comment]
Few more thoughts.

Answering my own question:
> Was that a correct decision back then?

**No, it was not.** Because people cannot pass objects to factories anymore. :( Well, they can, but the object will be cloned entirely, which most of the people don't want. But, if someone want, he can do it manually `stamp(clone(object))`.

I'd highly recommend us to make stampit v2 shallow copy it's argument. This is what people using ES6 would want because of new argument/object destructuring syntax.
`stamp({ database, userName, age: 42 })`

I'm starting the coding.

**Comment by lecstor** [comment]
for my 2c worth I ++ @koresar there. I can see benefit of "safety first" and protecting from effects from afar but I think most would expect references to remain intact. A convenience method for cloning wouldn't be a bad thing, maybe even if create() did the deep copy while stamp() remained shallow. Create kinda hints at creating a new thing. On the other hand, the simplicity of cloning before handing over the object as @koresar said negates the necessity of the deep clone option.

If I had gone the path I did with a little less stubbornness I likely would have bailed on stampit earlier without questioning it, accepting it wasn't the tool for my task at hand.

**Comment by troutowicz** [comment]
I guess I am totally lost now as to why no one else is seeing what I am seeing. @lecstor is not using stamps correctly at all. He doesn't understand the difference between a factory and a stamp instance and doesnt understand the behavior of `static` methods.

AFTER you call stampit(), ANOTHER factory is returned.

```js
let Node = stampit() // <---- Node is a factory NOT a stamp instance
  .refs({})
  .methods({});
```

At this point the contents of `refs` and `methods` are stored internally by the stamp factory, they CANNOT be used. Before trying to access or mutate state or call stamp methods, `Node` must stamp out a stamp instance!

```js
let nodeInstance = Node({ any refs to override });

nodeInstance.someRef;
nodeInstance.someMethod;
```

Everything @koresar has  said though still stands. The behavior of stamp factory args may need to be rethought.

**Comment by lecstor** [comment]
You did notice that I was calling the function returned by refs?
```
this.refs(args)()  <-
```
so the same as
```
newStamp = stampit.refs()
node = newStamp()
``
and the issue is that
```
let nodeInstance = Node({ any refs to override });  <- these "refs" are deepcopied
```
the code above that I posted runs as-is. You can copy/paste to a file and `babel-node` it.

**Comment by troutowicz** [comment]
Wow.... I apologize @lecstor! I never noticed the extra pair of parenthesis... my last couple comments can be scratched out and mostly forgot about.

What I now do not understand is why you think you need the static convenience method? What is wrong with

```
import stampit from 'stampit'
import map from 'lodash/collection/map'

let Node = stampit({
  methods: {
    setPaddingLeft(size) {
      this.attr.padding.left = size
    }
  },
})

let doc = { kids: [
  {attr: {padding: { left: 5, right: 10 } } },
  {attr: {padding: { left: 20, right: 20 } } }
]}

let node1 = Node.refs(doc.kids[0])();
let node2 = Node.refs(doc.kids[1])();
node1.setPaddingLeft(18)
node2.setPaddingLeft(16)
console.log('doc', JSON.stringify(doc, null, 2))
```

**Comment by lecstor** [comment]
8) no worries @troutowiz

No reason other than it bugs me. Both do exactly the same thing, one's just prettier in use.

Lecstor out.

**Comment by troutowicz** [comment]
I guess its your way of working around `create()` unexpected behavior. Makes sense.

Keep checking in here @lecstor as you have found a possible bug as @koresar mentioned.

**Comment by unstoppablecarl** [comment]
@koresar we talked about it here too #57

**Comment by koresar** [comment]
Thanks Carl!
Clearly, it's not even a bug but mis-implementation.
I beg your pardon, people. My fault. I forgot what we agreed to at the #57.

Good we haven't released the current master branch. ☺

If anyone could edit the README to point out that stamps accept refs as an
argument, that would be great. Thank you.

**Comment by ericelliott** [comment]
Just to clarify: invoking the stamp and calling `.create()` should do the same thing. Want a deep clone method? Call it something else.

**Comment by koresar** [comment]
PR fix #96 submitted. Please review.

**Comment by troutowicz** [comment]
Closing this, feel free to reopen if the PR did not satisfy you.
